File: apps/ios/BitOS/DesignSystem/SecretKeyField.swift
import BusinessCore
import SwiftUI
import UIKit


/**
 * Secret-key login field (ID-004), shared by the onboarding import step, the
 * You-tab import panel and the More-hub add-account sheet: masked by default
 * with reveal + one-tap paste, keyboard flags that never mangle a pasted
 * key, and live feedback from the shared `KeyImportForm` rule (verbatim copy
 * parity with Compose) rendered as an icon+text state banner.
 */
struct SecretKeyField: View {
    @Binding var text: String
    var error: String?
    var onSubmit: () -> Void = {}

    @State private var revealed = false

    private var check: BusinessCoreBridge.KeyImportCheckWire {
        BusinessCoreBridge().keyImportCheck(raw: text)
    }

    private var ready: Bool { check.verdict == "READY" }
    /// A submit error outranks the live hint; both clear on the next edit.
    private var feedback: String? { error ?? check.message }

    var body: some View {
        VStack(alignment: .leading, spacing: 8) {
            HStack(spacing: 0) {
                Group {
                    if revealed {
                        TextField("nsec1…", text: $text)
                    } else {
                        SecureField("nsec1…", text: $text)
                    }
                }
                .font(.system(size: 14, design: .monospaced))
                .foregroundStyle(BitOSTheme.textPrimary)
                .autocorrectionDisabled()
                .textInputAutocapitalization(.never)
                .keyboardType(.asciiCapable)
                .submitLabel(.continue)
                .onSubmit(onSubmit)
                .accessibilityLabel("Secret key")
                if text.isEmpty {
                    pasteButton
                }
                revealButton
            }
            .padding(.horizontal, 14)
            .padding(.vertical, 10)
            .background(
                RoundedRectangle(cornerRadius: 12, style: .continuous)
                    .fill(BitOSTheme.surfaceElevated)
            )
            .overlay(
                RoundedRectangle(cornerRadius: 12, style: .continuous)
                    .strokeBorder(borderColor, lineWidth: 1)
            )
            if let feedback {
                StateBannerView(
                    tone: ready ? .ok : .error,
                    text: feedback
                )
            }
        }
        .animation(.easeOut(duration: 0.15), value: feedback)
    }

    private var borderColor: Color {
        if ready { return BitOSTheme.success }
        if feedback != nil { return BitOSTheme.error }
        return BitOSTheme.divider
    }

    private var pasteButton: some View {
        Button {
            if let pasted = UIPasteboard.general.string {
                text = pasted
            }
        } label: {
            Text("Paste")
                .font(.system(size: 11, weight: .semibold))
                .foregroundStyle(BitOSTheme.textSecondary)
                .padding(.horizontal, 10)
                .padding(.vertical, 5)
                .background(Capsule().fill(BitOSTheme.surfaceOverlay))
                .overlay(Capsule().strokeBorder(BitOSTheme.border, lineWidth: 1))
                .padding(.trailing, 6)
        }
        .buttonStyle(.plain)
        .accessibilityLabel("Paste key from clipboard")
    }

    private var revealButton: some View {
        Button {
            revealed.toggle()
        } label: {
            Image(systemName: revealed ? "eye.slash" : "eye")
                .font(.system(size: 15))
                .foregroundStyle(BitOSTheme.textSecondary)
                .frame(width: 34, height: 34)
                .contentShape(Rectangle())
        }
        .buttonStyle(.plain)
        .accessibilityLabel(revealed ? "Hide key" : "Show key")
    }
}

/**
 * Live derived-identity preview (KF-6): once the shared rule resolves the
 * input to a usable secret, this shows the account the key controls — hex
 * avatar over the derived pubkey, monospace npub and a copy chip — before
 * anything is stored.
 */
struct DerivedIdentityCard: View {
    let check: BusinessCoreBridge.KeyImportCheckWire

    var body: some View {
        HStack(spacing: 12) {
            if let pubkey = check.pubkeyHex {
                HexAvatarView(pubkey: pubkey, size: 40)
            }
            VStack(alignment: .leading, spacing: 2) {
                Text("Derived identity")
                    .font(.system(size: 14, weight: .bold))
                    .foregroundStyle(BitOSTheme.textPrimary)
                if let npub = check.npub {
                    Text(middleEllipsized(npub))
                        .font(.system(size: 11, design: .monospaced))
                        .foregroundStyle(BitOSTheme.textSecondary)
                        .lineLimit(1)
                }
            }
            .frame(maxWidth: .infinity, alignment: .leading)
            if let npub = check.npub {
                Button {
                    UIPasteboard.general.string = npub
                } label: {
                    Label("Copy npub", systemImage: "doc.on.doc")
                        .font(.system(size: 11, weight: .semibold))
                        .foregroundStyle(BitOSTheme.accent)
                        .padding(.horizontal, 10)
                        .padding(.vertical, 6)
                        .background(Capsule().fill(BitOSTheme.accent.opacity(0.14)))
                        .overlay(Capsule().strokeBorder(BitOSTheme.accent.opacity(0.4), lineWidth: 1))
                }
                .buttonStyle(.plain)
                .accessibilityLabel("Copy npub")
            }
        }
        .padding(14)
        .background(
            RoundedRectangle(cornerRadius: 14, style: .continuous)
                .fill(BitOSTheme.surfaceElevated)
        )
        .overlay(
            RoundedRectangle(cornerRadius: 14, style: .continuous)
                .strokeBorder(BitOSTheme.border, lineWidth: 1)
        )
    }
}

/** True when the field text resolves to a usable secret key. */
func secretKeyReady(_ text: String) -> Bool {
    BusinessCoreBridge().keyImportCheck(raw: text).verdict == "READY"
}

File: apps/android/src/main/kotlin/space/bitos/app/ui/components/SecretKeyField.kt
package space.bitos.app.ui.components

import androidx.compose.foundation.background
import androidx.compose.foundation.border
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.foundation.text.KeyboardActions
import androidx.compose.foundation.text.KeyboardOptions
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.outlined.ContentCopy
import androidx.compose.material.icons.outlined.Visibility
import androidx.compose.material.icons.outlined.VisibilityOff
import androidx.compose.material3.Icon
import androidx.compose.material3.IconButton
import androidx.compose.material3.OutlinedTextField
import androidx.compose.material3.OutlinedTextFieldDefaults
import androidx.compose.material3.Surface
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.platform.LocalClipboardManager
import androidx.compose.ui.semantics.contentDescription
import androidx.compose.ui.semantics.semantics
import androidx.compose.ui.text.AnnotatedString
import androidx.compose.ui.text.TextStyle
import androidx.compose.ui.text.font.FontFamily
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.input.ImeAction
import androidx.compose.ui.text.input.KeyboardCapitalization
import androidx.compose.ui.text.input.KeyboardType
import androidx.compose.ui.text.input.PasswordVisualTransformation
import androidx.compose.ui.text.input.VisualTransformation
import androidx.compose.ui.text.style.TextOverflow
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import space.bitos.app.ui.theme.BitOSColors
import space.bitos.core.identity.KeyImportCheck
import space.bitos.core.identity.KeyImportForm
import space.bitos.core.identity.KeyImportVerdict

/**
 * Secret-key login field (ID-004), shared by the onboarding import step, the
 * You-tab import panel and the More-hub add-account sheet: masked by default
 * with reveal + one-tap paste, keyboard flags that never mangle a pasted
 * key, and live feedback from the shared `KeyImportForm` rule rendered as an
 * icon+text state banner. The caller owns the text; gate the submit action
 * on `KeyImportForm.check(value).verdict == READY`.
 */
@Composable
fun SecretKeyField(
    value: String,
    onValueChange: (String) -> Unit,
    error: String?,
    modifier: Modifier = Modifier,
    onSubmit: () -> Unit = {},
    textStyle: TextStyle = TextStyle.Default,
) {
    var revealed by remember { mutableStateOf(false) }
    val clipboard = LocalClipboardManager.current
    val check = remember(value) { KeyImportForm.check(value) }
    val ready = check.verdict == KeyImportVerdict.READY
    // A submit error outranks the live hint; both clear on the next edit.
    val feedback = error ?: check.message

    Column(modifier = modifier, verticalArrangement = Arrangement.spacedBy(8.dp)) {
        OutlinedTextField(
            value = value,
            onValueChange = onValueChange,
            placeholder = {
                Text("nsec1…", fontFamily = FontFamily.Monospace, color = BitOSColors.textTertiary)
            },
            visualTransformation = if (revealed) VisualTransformation.None else PasswordVisualTransformation(),
            singleLine = true,
            textStyle = textStyle.copy(fontFamily = FontFamily.Monospace),
            shape = RoundedCornerShape(12.dp),
            colors = OutlinedTextFieldDefaults.colors(
                focusedBorderColor = if (ready) BitOSColors.success else BitOSColors.primary,
                unfocusedBorderColor = if (feedback != null && !ready) BitOSColors.error else BitOSColors.border,
                focusedContainerColor = BitOSColors.surfaceElevated,
                unfocusedContainerColor = BitOSColors.surfaceElevated,
                cursorColor = BitOSColors.primary,
            ),
            trailingIcon = {
                Row(verticalAlignment = Alignment.CenterVertically) {
                    if (value.isEmpty()) {
                        Surface(
                            shape = RoundedCornerShape(999.dp),
                            color = BitOSColors.surfaceOverlay,
                            border = androidx.compose.foundation.BorderStroke(1.dp, BitOSColors.border),
                            modifier = Modifier
                                .padding(end = 4.dp)
                                .clickable(onClickLabel = "Paste key from clipboard") {
                                    clipboard.getText()?.text?.let(onValueChange)
                                },
                        ) {
                            Text(
                                "Paste",
                                color = BitOSColors.textSecondary,
                                fontSize = 11.sp,
                                fontWeight = FontWeight.W600,
                                modifier = Modifier.padding(horizontal = 10.dp, vertical = 5.dp),
                            )
                        }
                    }
                    IconButton(onClick = { revealed = !revealed }) {
                        Icon(
                            if (revealed) Icons.Outlined.VisibilityOff else Icons.Outlined.Visibility,
                            contentDescription = if (revealed) "Hide key" else "Show key",
                            tint = BitOSColors.textSecondary,
                        )
                    }
                }
            },
            keyboardOptions = KeyboardOptions(
                keyboardType = KeyboardType.Password,
                autoCorrect = false,
                capitalization = KeyboardCapitalization.None,
                imeAction = ImeAction.Done,
            ),
            keyboardActions = KeyboardActions(onDone = { if (ready) onSubmit() }),
            modifier = Modifier
                .fillMaxWidth()
                .semantics { contentDescription = "Secret key" },
        )
        feedback?.let { message ->
            StateBanner(
                tone = if (ready) StateBannerTone.OK else StateBannerTone.ERROR,
                text = message,
            )
        }
    }
}

/**
 * Live derived-identity preview (KF-6): once the shared rule resolves the
 * input to a usable secret, this shows the account the key controls — hex
 * avatar over the derived pubkey, monospace npub and a copy chip — before
 * anything is stored.
 */
@Composable
fun DerivedIdentityCard(
    check: KeyImportCheck,
    modifier: Modifier = Modifier,
) {
    val pubkey = check.pubkeyHex ?: return
    val npub = check.npub ?: return
    val clipboard = LocalClipboardManager.current
    Surface(
        shape = RoundedCornerShape(14.dp),
        color = BitOSColors.surfaceElevated,
        border = androidx.compose.foundation.BorderStroke(1.dp, BitOSColors.border),
        modifier = modifier,
    ) {
        Row(
            verticalAlignment = Alignment.CenterVertically,
            horizontalArrangement = Arrangement.spacedBy(12.dp),
            modifier = Modifier.padding(14.dp),
        ) {
            HexAvatar(pubkey = pubkey, size = 40)
            Column(Modifier.weight(1f)) {
                Text(
                    "Derived identity",
                    fontSize = 14.sp,
                    fontWeight = FontWeight.W700,
                    color = BitOSColors.textPrimary,
                )
                Text(
                    middleEllipsize(npub),
                    fontSize = 11.sp,
                    fontFamily = FontFamily.Monospace,
                    color = BitOSColors.textSecondary,
                    maxLines = 1,
                    overflow = TextOverflow.Ellipsis,
                )
            }
            Surface(
                shape = RoundedCornerShape(999.dp),
                color = BitOSColors.primary.copy(alpha = 0.14f),
                border = androidx.compose.foundation.BorderStroke(1.dp, BitOSColors.primary.copy(alpha = 0.4f)),
                modifier = Modifier.clickable(onClickLabel = "Copy npub") {
                    clipboard.setText(AnnotatedString(npub))
                },
            ) {
                Row(
                    verticalAlignment = Alignment.CenterVertically,
                    horizontalArrangement = Arrangement.spacedBy(4.dp),
                    modifier = Modifier.padding(horizontal = 10.dp, vertical = 5.dp),
                ) {
                    Icon(
                        Icons.Outlined.ContentCopy,
                        contentDescription = null,
                        tint = BitOSColors.primary,
                        modifier = Modifier.size(12.dp),
                    )
                    Text(
                        "Copy npub",
                        color = BitOSColors.primary,
                        fontSize = 11.sp,
                        fontWeight = FontWeight.W600,
                    )
                }
            }
        }
    }
}

/** True when the current field text resolves to a usable secret key. */
fun secretKeyReady(value: String): Boolean = KeyImportForm.check(value).verdict == KeyImportVerdict.READY

/** npub/nsec display truncation (CB-1): keep both bech32 ends visible. */
fun middleEllipsize(value: String, head: Int = 10, tail: Int = 8): String =
    if (value.length <= head + tail + 1) value else value.take(head) + "…" + value.takeLast(tail)
